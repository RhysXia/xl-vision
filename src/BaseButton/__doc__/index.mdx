# BaseButton

基础的 button 组件，只提供了基础的功能，可以方便使用者自定义。

## 代码演示

:::demo 自定义按钮

自定义按钮

```jsx
import React from 'react'
import { BaseButton } from 'xl-vision'

export default () => {
  return (
    <div className='container'>
      <BaseButton className='custom'>custom css</BaseButton>
    </div>
  )
}
```

```scss
@use '../style/var' as *;

.container {
  &::deep {
    .custom {
      padding: 6px 16px;
      border-radius: 3px;
      background-color: #9c27b0;
      color: #fff;

      @include button;
      @include elevation(2);

      &:focus {
        opacity: 0.85;
      }
    }
  }
}
```

:::

:::demo 复杂按钮

复杂按钮

```jsx
import React from 'react'
import { BaseButton } from 'xl-vision'
import Breakfast from './breakfast.jpg'
import Burgers from './burgers.jpg'
import Camera from './camera.jpg'
const images = [
  {
    url: Breakfast,
    title: 'Breakfast',
    width: '40%'
  },
  {
    url: Burgers,
    title: 'Burgers',
    width: '30%'
  },
  {
    url: Camera,
    title: 'Camera',
    width: '30%'
  }
]

export default () => {
  return (
    <div className='container'>
      {images.map((it, index) => (
        <BaseButton
          key={index}
          style={{
            width: it.width
          }}
          className='button'
        >
          <span
            className='img'
            style={{
              backgroundImage: `url(${it.url})`
            }}
          />
          <span className='title'>{it.title}</span>
        </BaseButton>
      ))}
    </div>
  )
}
```

```scss
.container {
  &::deep {
    .button {
      position: relative;
      height: 200px;
      color: #000;
      &:focus {
        opacity: 0.85;
      }
    }
    .img {
      display: block;
      position: absolute;
      left: 0;
      top: 0;
      right: 0;
      bottom: 0;
      background-size: cover;
    }
    .title {
      position: absolute;
      left: 50%;
      top: 50%;
      transform: translate(-50%, -50%);
      color: #fff;
      font-size: 20px;
      font-weight: bold;
    }
  }
}
```

:::

## BaseButton API

| 参数          | 说明               | 类型      | 可选值 | 默认值 | 是否必填 |
| ------------- | ------------------ | --------- | ------ | ------ | -------- |
| clsPrefix     | 自定义类名前缀     | string    | -      | -      | false    |
| disableRipple | 是否禁止水波纹特效 | boolean   | -      | true   | false    |
| disabled      | 是否开启禁止状态   | boolean   | -      | false  | false    |
| loading       | 是否开启加载状态   | boolean   | -      | false  | false    |
| href          | 指定点击跳转的链接 | string    | -      | -      | false    |
| children      | 子节点             | ReactNode | -      | -      | true     |
