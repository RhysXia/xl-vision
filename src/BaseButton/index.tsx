import React from 'react'
import Ripple from '../Ripple'
import ConfigContext from '../ConfigProvider/ConfigContext'
import classnames from 'classnames'
import PropTypes from 'prop-types'

export interface CommonBaseButtonProps {
  href?: string
  enableRipple?: boolean
  clsPrefix?: string
  disabled?: boolean
  loading?: boolean
  onClick?: React.EventHandler<React.MouseEvent>
}

export type BaseButtonProps = CommonBaseButtonProps &
  React.ButtonHTMLAttributes<HTMLButtonElement> &
  React.AnchorHTMLAttributes<HTMLAnchorElement>

export type ButtonElement = HTMLButtonElement & HTMLAnchorElement

const BaseButton = React.forwardRef<ButtonElement, BaseButtonProps>((props, ref) => {
  const { clsPrefix: rootClsPrefix } = React.useContext(ConfigContext)

  const {
    clsPrefix = `${rootClsPrefix}-base-button`,
    children,
    enableRipple = true,
    href,
    disabled,
    loading,
    onClick,
    className,
    ...others
  } = props

  const Component = href ? 'a' : 'button'

  const classes = classnames(
    clsPrefix,
    {
      [`${clsPrefix}--disabled`]: disabled,
      [`${clsPrefix}--loading`]: loading && !disabled
    },
    className
  )

  const onClickWrap = React.useCallback(
    (e: React.MouseEvent) => {
      if (loading || disabled) {
        e.preventDefault()
        return
      }
      onClick && onClick(e)
    },
    [loading, disabled, onClick]
  )

  const shouldEnableRipple = enableRipple && !disabled && !loading

  return (
    <Component {...others} onClick={onClickWrap} className={classes} ref={ref}>
      <span>{children}</span>
      <Ripple
        disableEvents={!shouldEnableRipple}
        leaveAfterEnter={true}
        transitionClasses={`${clsPrefix}__ripple`}
      />
    </Component>
  )
})

BaseButton.displayName = 'BaseButton'

BaseButton.propTypes = {
  clsPrefix: PropTypes.string,
  children: PropTypes.node,
  enableRipple: PropTypes.bool,
  href: PropTypes.string,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  onClick: PropTypes.func,
  className: PropTypes.string
}

export default BaseButton
