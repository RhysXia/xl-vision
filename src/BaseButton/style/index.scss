@use './var' as *;
@use '../../Ripple/style';

$button: #{$cls-prefix}-base-button;
$ripple: #{$cls-prefix}-ripple;

.#{$button} {
  position: relative;
  display: inline-block;
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  white-space: nowrap;
  text-align: center;
  text-decoration: none;
  vertical-align: middle;
  background-color: transparent;
  border: 0;
  border-radius: 0;
  outline: none;
  cursor: pointer;
  -webkit-appearance: none;
  user-select: none;
  touch-action: manipulation;
  // 阻止在移动端按下按钮后闪烁的问题
  -webkit-tap-highlight-color: transparent;

  &--disabled {
    cursor: not-allowed;
  }

  &--loading {
    pointer-events: none;
  }

  &__inner {
    // 阻止ie下 focus时文字移动
    position: relative;
  }

  .#{$ripple} {
    transform: scale(1);
    opacity: $base-button-ripple-opacity-pressed;
  }

  &__ripple {
    &-enter-active,
    &-leave-active {
      transition: $base-button-ripple-transition;
    }

    &-enter {
      transform: scale(0);
      opacity: $base-button-ripple-opacity-enter;
    }

    &-leave-to {
      opacity: $base-button-ripple-opacity-enter;
    }
  }
}
