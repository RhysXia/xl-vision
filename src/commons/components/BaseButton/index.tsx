import React from 'react'

export type NativeButtonProps = Omit<React.AnchorHTMLAttributes<HTMLAnchorElement>, 'type'> &
  Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'type'>

export interface BaseButtonProps extends NativeButtonProps {
  children: React.ReactNode
  className?: string
  clsPrefix?: string
  component?: React.ElementType
  disabled?: boolean
}

const BaseButton: React.ForwardRefRenderFunction<any, BaseButtonProps> = (props, ref) => {
  const {
    children,
    className,
    clsPrefix,
    component = 'button',
    disabled = false,
    ...others
  } = props

  return <div></div>
}

export default BaseButton
