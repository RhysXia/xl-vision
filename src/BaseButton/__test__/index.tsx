import React from 'react'
import { mount } from 'enzyme'
import BaseButton from '..'

describe('BaseButton', () => {
  it('基本测试', () => {
    const handleClick = jest.fn()

    const wrapper = mount(<BaseButton onClick={handleClick}>button</BaseButton>)

    wrapper.find('.xl-base-button').simulate('click')
    wrapper.update()

    expect(handleClick.mock.calls.length).toBe(1)

    expect(wrapper.render()).toMatchSnapshot()
  })

  it('测试disabled', () => {
    const handleClick = jest.fn()

    const wrapper = mount(
      <BaseButton disabled={true} onClick={handleClick}>
        button
      </BaseButton>
    )

    wrapper.find('.xl-base-button').simulate('click')
    wrapper.update()

    expect(handleClick.mock.calls.length).toBe(0)
  })

  it('测试loading', () => {
    const handleClick = jest.fn()

    const wrapper = mount(
      <BaseButton loading={true} onClick={handleClick}>
        button
      </BaseButton>
    )

    wrapper.find('.xl-base-button').simulate('click')
    wrapper.update()

    expect(handleClick.mock.calls.length).toBe(0)
  })
})
