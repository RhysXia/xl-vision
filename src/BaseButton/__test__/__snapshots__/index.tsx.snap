// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseButton 基本测试 1`] = `
<button
  class="xl-base-button"
  tabindex="0"
>
  <span
    class="xl-base-button__inner"
  >
    button
  </span>
  <span
    class="xl-ripple"
  />
</button>
`;
